### PR TITLE
WIP: Replace SpanInfo with HIE Files

### PR DESCRIPTION
**Pull request by wz1000**
Removes all traces of `SpanInfo`, adds HIE file support for 8.6, interns in a version of `HieBin` that supports updating the `NameCache`.

New features:

- Document Highlight

Re-implemented features:

- Hover
- Go to (type)definition

Missing features:

- The "relevant constraint" on hover stuff. 
   I will add this once @mpickering approves my typeclass info GHC MR.

Fixed:

- `readHieFile` uses and updates the global IDE `NameCache` atomically

Internal changes:

- We now have a global `NameCache`
- HIE file generation for 8.6, adapted from 8.8
- `HieBin` module copied in for all versions of GHC
- Add `useWithStaleFast'` to both return a stale result as well as a barrier to wait on the up-to-date result
- New Rules:
  + `GetHieFile` - self explanatory - potential for improvement
  + `GetRefMap` - Generate a references map from the HIE file - used for document highlight
  + `GetDocMap` - Generate a map from names to documentation - used for hover

TODO:

- [ ] Make hie/hi file generation more robust
- [x] ~Investigate if we really need to load ifaces for all downstream modules in order to generate the `DocMap`~ Guess we do

### CODE REVIEW

File: exe/Main.hs [hunk 1]
```diff
@@ -161,7 +163,9 @@ main = do
         putStrLn "\nStep 3/6: Initializing the IDE"
         vfs <- makeVFSHandle
         debouncer <- newAsyncDebouncer
-        (ide, worker) <- initialise def mainRule (pure $ IdInt 0) (showEvent lock) (logger Debug) debouncer (defaultIdeOptions $ loadSession dir) vfs
+        us      <- mkSplitUniqSupply 'r'
+        nc_var  <- newIORef (initNameCache us knownKeyNames)
+        (ide, worker) <- initialise def mainRule (pure $ IdInt 0) (showEvent lock) (logger Debug) debouncer nc_var (defaultIdeOptions $ loadSession dir) vfs
```
Comment (mpickering) [inline exe/Main.hs]: I pushed this stuff into `shakeOpen`.

File: src/Development/IDE/Core/RuleTypes.hs [hunk 2]
```diff
@@ -52,6 +52,7 @@ type instance RuleResult GetModuleGraph = DependencyInformation
 data TcModuleResult = TcModuleResult
     { tmrModule     :: TypecheckedModule
     , tmrModInfo    :: HomeModInfo
+    , tmrHieFile    :: Maybe HieFile
```
Comment (mpickering) [inline src/Development/IDE/Core/RuleTypes.hs:55]: Didn't we discuss this was inconsistent with how all the other artifacts are generated?

File: src/Development/IDE/Core/Rules.hs [hunk 3]
```diff
@@ -121,25 +121,34 @@ getAtPoint :: NormalizedFilePath -> Position -> IdeAction (Maybe (Maybe Range, [
 getAtPoint file pos = fmap join $ runMaybeT $ do
   ide <- ask
   opts <- liftIO $ getIdeOptionsIO ide
-  (spans, mapping) <- useE  GetSpanInfo file
+
+  (hf, mapping) <- useE GetHieFile file
+  (PDocMap dm,_) <- useE GetDocMap file
```
Comment (mpickering) [inline src/Development/IDE/Core/Rules.hs:126]: The `DocMap` could be at a different version to the `HieFile`? Needs to be justified with a comment which it is correct to ignore the position mapping.

File: src/Development/IDE/Core/Rules.hs [hunk 4]
```diff
@@ -156,32 +165,40 @@ getHieFile ide file mod = do
...
-      lift $ delayedAction (mkDelayedAction "OutOfDateHie" ("hie" :: T.Text, f) L.Info
-                              (do pm <- use_ GetParsedModule f
-                                  typeCheckRuleDefinition f pm DoGenerateInterfaceFiles
-                                  liftIO $ signalBarrier b ()))
-      () <- MaybeT $ liftIO $ timeout 1 $ waitBarrier b
-      liftIO $ loadHieFile hie_f
+  hfr <- lift $ useWithStaleFast' GetHieFile f
+  case stale hfr of
+    Just (hf,_) -> pure hf -- We already have the file
+    Nothing -> do -- We don't have the file, so try loading it from disk
```
Comment (mpickering) [inline src/Development/IDE/Core/Rules.hs:178]: We discussed why this doesn't just block on the call to `useWithStaleFast` from above. Needs a comment explaining what the logic behind this code is.

**Comment by mpickering** [comment]
Overall looks really good. Definitely needs comments clarify what is going on in `getHomeHieFile`. Please can you also split it up into somewhat logical commits? For example:

* Fix warnings
* Add COMPLETE pragma
* Adding `useWithStale'`
* Add HIE file for 8.6
* Fix the namecache reading for HIE files
* Add the HIE file rule
* Add the doc map rule
* Reimplement hover etc with HIE file rule
* Make FileStore use GetHieFile
* Implement document symbols

Tested so far on ghicde and generic-lens, going to test on GHC now.

File: src/Development/IDE/Core/RuleTypes.hs (hunk 2 above)
Comment (wz1000) [reply src/Development/IDE/Core/RuleTypes.hs:55]: Yeah, what I got from the conversation was that maybe handling `.hi` files in the same way would make sense. The TODO in description partly refers to this.

File: src/Development/IDE/Core/Rules.hs (hunk 3 above)
Comment (wz1000) [reply src/Development/IDE/Core/Rules.hs:126]: I don't think that's intended. They should be the same version. Is there a way to enforce this? The `DocMap` rule depends on the `GetHieFile` rule.

**Pull request closed without merging**